common/src/main/scala/com/gu/recipeasy/models/RecipeReadiness.scala

```diff
@@ -11,17 +19,142 @@ object RecipeReadiness {
+    def text(node: Node): String = node match {
+      case tn: TextNode => tn.text.trim
+      case elem: Element => elem.text.trim
+      case other => ""
+    }
+    val listItems: Seq[Node] = para.childNodes.asScala.filterNot(_.nodeName == "br")
+    val withoutServingCount = listItems.filterNot(n => text(n).toLowerCase.dropWhile(_ == '(').startsWith("serves"))
+    if (withoutServingCount.size < 2) None
+    else {
+      if (withoutServingCount(0).nodeName != withoutServingCount(1).nodeName) {
+        val title = text(withoutServingCount(0))
+        val ingredients = withoutServingCount.drop(1).map(text)
+        Some(IngredientsList(title = Some(title), ingredients))
+      } else {
+        val ingredients = withoutServingCount.map(text)
+        Some(IngredientsList(title = None, ingredients))
+      }
+    }
+  }
+  def guessIngredients(body: Seq[Element]): Seq[IngredientsList] = {
+    // Find paragraphs containing short text items separated by <br>
+    val candidates = body.filter(ParaWithListOfShortTexts.matches)
+    candidates.flatMap(buildIngredientList)
+  }
+
+  def extractIngredientsLists(recipe: Recipe): Seq[IngredientsList] = {
+    val doc = Jsoup.parse(recipe.body)
+    val body: Seq[Element] = doc.body.children.asScala
+    val ingredientsLists: Seq[IngredientsList] = guessIngredients(body)
+    ingredientsLists
+  }
+
+  def computeUpdatedCuratedRecipe(db: DB, recipe: Recipe, newIngredientsLists: Seq[IngredientsList]): Option[CuratedRecipe] = {
+    val maybeCuratedRecipe_Old: Option[CuratedRecipe] = db.getOriginalRecipe(recipe.id).map(CuratedRecipe.fromRecipe(_))
+    maybeCuratedRecipe_Old match {
+      case None => None
+      case Some(curatedRecipeOld) => {
+        val seqDetailedIngredientsList: Seq[DetailedIngredientsList] = newIngredientsLists.map { isl =>
+          DetailedIngredientsList(
+            isl.title,
+            isl.ingredients.map(string => DetailedIngredient(
+              Some(1.toDouble),
+              None,
+              string,
+              None,
+              string
+            ))
+          )
+        }
+        val detailedIngredientsLists: DetailedIngredientsLists = DetailedIngredientsLists(seqDetailedIngredientsList)
+        val curatedRecipeNew = curatedRecipeOld.copy(
+          ingredientsLists = detailedIngredientsLists
+        )
+        Some(curatedRecipeNew)
+      }
+    }
+  }
+
+  // -------------------------------------------------
+  // Stage 3
+
+  def reParseNewRecipesWithEmptyIngredientLists(db: DB): Int = {
+    val recipes: List[Recipe] = db.getOriginalRecipes()
+      .filter(_.status == New)
+      .filter(_.ingredientsLists.lists.size == 0) // selecting the ones with empty ingredient lists
+      .filter(recipe => !recipeHasACuratedRecordWithNonEmptyIngredientsLists(db, recipe)) // selecting the one which have not been acted on yet
+      .filter(recipe => extractIngredientsLists(recipe).size > 0)
```

Comment:
`extractIngredientsLists` looks quite expensive, can we lose this `filter` and just add an `if` statement in the `foreach` below?